src/acl/FilledChecklist.cc

```diff
@@ -247,9 +249,36 @@ ACLFilledChecklist::ACLFilledChecklist(const acl_access *A, HttpRequest *http_re
     setIdent(ident);
 }
 
+void
+ACLFilledChecklist::setPeer(const CachePeer * const peer)
+{
+    // TODO: Replace code that still does this manually with setPeer().
+    if (dst_peer_name.isEmpty() && peer)
+        dst_peer_name = peer->name;
+}
+
+void
+ACLFilledChecklist::setOutgoingConnection(const Comm::ConnectionPointer &outgoingConn)
+{
+    // TODO: Replace code that still does this manually with setOutgoingConnection().
+    if (!outgoingConn)
+        return;
+
+    setPeer(outgoingConn->getPeer());
+
+    // TODO: ACLIP::match(ip) does not check whether ip has been set. Should it?
```

Comment:
It feels wrong that ACLIP::match() was (and still is) treating default-constructed (i.e. setEmpty()) IP addresses as regular IP addresses. Most likely, those checks never actually match any IP an admin can configure via squid.conf, so I doubt there is a big problem with that, but it is still conceptually wrong to silently use an unknown/unset address IMO. I will remove this TODO without arguing if it turns out to be controversial.

Reply:
FTR, setEmpty() initializes to IPv6 ANY_ADDR which is valid for display and matches as non-IPv4 (eg ```http_access deny !ipv4```) while being invalid for routing. It can be configured in any IP address type ACL with ```::``` (or equivalent longer strings).

Reply:
There is a ton of opportunities for a serious misunderstanding here! I will highlight a few relevant facts in hope to reduce the number of such opportunities.

Ip::Address is more than just an IN6/sin6_addr address. It also manages address family, address port, and other address info:

```C++
// Ip::Address manages the whole sockaddr_in6 rather than just sin6_addr:
struct sockaddr_in6 {
    sa_family_t     sin6_family;
    in_port_t       sin6_port;
    uint32_t        sin6_flowinfo;
    struct in6_addr sin6_addr;
    uint32_t        sin6_scope_id;
};
```

Ip::Address::setEmpty() zeros the _whole_ Ip::Address object. There is currently no single Ip::Address method that identifies such objects except for the compareWhole() method with another setEmpty() address as a parameter. If we create such a method and call it isEmpty() then, for some Ip::Address objects, isEmpty() will not return the same value as IsAny(). Those are two _different_ concepts/states. For convenience, I will use isEmpty() as a short version of "compareWhole() method with another setEmpty() address as a parameter".

Calling setEmpty() does make Ip::Address::isAny() true (because IN6_IS_ADDR_UNSPECIFIED() returns true for an all-zero IN6 address), but that is _not_ the right way to make that method true. The right way to make isAny() true is to call setAnyAddr().

Configuring an IP ACL with a `::` address creates an acl_ip_data object rather than (just) an Ip::Address object. The first Ip::Address part of that configured acl_ip_data object (i.e. acl_ip_data::addr1) is _not_ all-zeros. The address family of add1 is set to 10 when the configuration says `::`. Thus, that addr1 isAny() but not isEmpty(). There might be a way to configure an IP ACL to have an all-zeros acl_ip_data::addr1, but I do not know it.

The TODO in question is _not_ about addresses used to configure IP ACLs but about traffic IP addresses being fed to the already configured IP ACLs. The current ACLIP::match() code does not ignore all-zeros (i.e. unknown/unset/isEmpty()) traffic addresses, and those addresses will match some of the configured ACL addresses, including `::`. This out-of-scope TODO should decide whether such ACL matches are a bug.